Setup i18n for the new request form

Fredx87 opened the pull request:
## Description

This PR adds internationalization for the New Request Form, using the [react-i18next](https://react.i18next.com/) library. Hardcoded strings are replaced with calls to the i18n library and some tooling has been added to load translations from the Zendesk CDN, to extract translation strings from the code, and to add linting rules to avoid hardcoded strings. It also moves a couple of elements (`Follow-up to request {{parentId}}` and `Fields marked with an asterisk (*) are required.`) from the Curlybars template to the JS component, to be translated in the same way.

### Translation loading

Translations can be downloaded from the Zendesk Static Assets CDN using the `bin/update-module-translations.mjs` script. The translation files are placed in the `src/modules/new-request-form/translations/locales` and are bundled by Rollup in a single bundle file.

This is not ideal because we are creating a big single bundle, but it is the best solution at the moment to avoid the creation of a lot of assets or depending on the URLs of the translations already published on the CDN.

The size of the single bundle with the translations is ~30 Kb when gzipped:
```sh
$ gzip assets/* && du -sh assets/*
 16K    assets/new-request-form-bundle.js.gz
 28K    assets/new-request-form-translations-bundle.js.gz
388K    assets/vendor-bundle.js.gz
```
### String extraction

We need to create a YML file containing the English strings which is picked up by our internal translation system. Developers can provide the English default value in the JS source code, and then run the `bin/extract-strings.mjs` to update the YML file, adding additional info (`screenshot` and `description`) manually. 

### Linting

I've added the `@shopify/jsx-no-hardcoded-content` rule from the `@shopify/eslint-plugin` to warn developers when hardcoded content is present in the code. This should be helpful to find when we are missing some translations, but it is not perfect and for this reason, it is reported as a warning and not an error.

![Screenshot 2024-04-02 at 15 48 20](https://github.com/zendesk/copenhagen_theme/assets/13420283/1647466f-3e2d-4346-b602-65c1f6bba0d5)

The rule is configurable and at the moment it will check for hardcoded content in component children, `aria-label` and `title` attributes, and the `content` prop of the Garden Tooltip component. It also excludes the `Trans` component from `react-i18next`. We can expand the rules in the future if we find that other checks are needed.

## Screenshots

Some screenshots of the New Request Form in English, Arabic, and Danish:

![Screenshot 2024-04-02 at 15 59 44](https://github.com/zendesk/copenhagen_theme/assets/13420283/ad42ce5a-1718-47f6-942a-6822c60d14f3)

![Screenshot 2024-04-02 at 16 01 15](https://github.com/zendesk/copenhagen_theme/assets/13420283/4460d346-2ecb-4e07-9b8c-b565bcee410c)

![Screenshot 2024-04-02 at 16 01 44](https://github.com/zendesk/copenhagen_theme/assets/13420283/caf02c7a-ea6c-4bfb-92be-124743125319)

## Checklist

- [x] :green_book: all commit messages follow the [conventional commits](https://conventionalcommits.org/) standard
- [x] :arrow_left: changes are compatible with RTL direction
- [x] :wheelchair: Changes to the UI are [tested for accessibility](./../README.md#accessibility-testing) and compliant with [WCAG 2.1](https://www.w3.org/TR/WCAG21/).
- [x] :memo: changes are tested in Chrome, Firefox, Safari and Edge
- [x] :iphone: changes are responsive and tested in mobile
- [ ] :+1: PR is approved by @zendesk/vikings

<!-- More info about the contribution process can be found at https://github.com/zendesk/copenhagen_theme#contributing -->

luis-almeida commented:
Is the list in `guide_available_locales.json` different than what we get in https://static.zdassets.com/translations/new-request-form/manifest.json?

Fredx87 commented:
> Is the list in `guide_available_locales.json` different than what we get in https://static.zdassets.com/translations/new-request-form/manifest.json?

Yes, I performed some checks on the two files to be sure that the implementation is right.

`guide_available_locales.json` contains 359 entries, while the manifest has 86 entries.

- A locale in `guide_available_locales.json` can have `"parent_identifier": null`. This means that is a standard locale and there is an entry in the manifest for it
  - I found an exception for this rule which is the `tl-ph` locale. I don't know if it is an error, but in this case, we would fallback to the English strings
- A locale in `guide_available_locales.json` can have a `parent_identifier`.  For example, the `nn` locale has `no` as a parent, and the `en-dk` has `en-150` as a parent.
  - In general, we can read the `parent_identifier` from the `guide_available_locales.json` and find the respective entry in the manifest
  - I just found out that in some cases there are multiple levels in the `parent_identifier`, for example, `no-nn` has `nn` as a parent which has `no` as a parent 🙈. This thing is not considered in the current implementation, because it will stop at `nn` and it will not find an entry in the manifest for it.

luis-almeida commented:
Thanks for explaining it in detail! ❤️ 

What I was trying to understand is if we really need to read from `guide_available_locales`.
In theory Help Center needs to support all available locales as there's no other product at Zendesk that supports more languages than Guide (HC).
So if the manifest is already listing all available translations, would it be enough to just use them all? And not even check `guide_available_locales`.

Fredx87 commented:
> So if the manifest is already listing all available translations, would it be enough to just use them all? And not even check `guide_available_locales`.

I don't understand how this should work. The manifest lists all available translations, but Help Center has more locales that are mapped to a single translation locale.

For example, if the Help Center locale is `nn` we cannot find the `nn` entry in the manifest. Using `guide_available_locales` we can check the parent id and get `no`, and at this point, we can find the entry in the manifest and fetch the right translation file. The same concept applies to 282 locates that have a `parent_identifier` in `guide_available_locales`. I cannot think of a solution that doesn't involve `guide_available_locales`.

As I said, exposing this complexity in the theme and maintaining it in the long term is not ideal. I think that the best solution at this point is to pass this info from Curlybars, for example adding a `cldr_locale` attribute to the `help_center` object, and letting the backend perform the mapping.

Fredx87 commented on package.json:
```diff
@@ -11,6 +11,8 @@
     "download-locales": "node ./bin/update-translations",
     "test": "jest",
     "test-a11y": "node bin/lighthouse/index.js",
+    "i18n:extract": "node bin/extract-strings.mjs new-request-form 'src/modules/new-request-form/**/*.{ts,tsx}' src/modules/new-request-form/translations",
+    "i18n:update-translations": "node bin/update-module-translations.mjs new-request-form src/modules/new-request-form/translations/locales",
```
At the moment these scripts are referring only to the `new-request-form`. The setup can be revisited later when we'll have more components

Fredx87 commented:
I updated the PR removing the mapping logic, and adding the download and bundling of the translation files

luis-almeida commented on README.md:
```diff
@@ -107,6 +107,54 @@ Notes:
+
+#### Updating translation files
+
+Use the `bin/update-module-translations.mjs` to download the latest translations for a module. All files are then bundled by the build process in a single `[MODULE]-translations-bundle.js` file.
```
Would it make sense to have it work for all "apps" and not just for a given module?
When updating translations we usually always want them all.

luis-almeida commented on bin/update-module-translations.mjs:
```diff
@@ -0,0 +1,32 @@
+import { writeFile } from "node:fs/promises";
+import { resolve } from "node:path";
+
+const PACKAGE_NAME = process.argv[2];
```
Same thing here as it could also apply to all modules.

luis-almeida commented on bin/update-module-translations.mjs:
```diff
@@ -0,0 +1,32 @@
+import { resolve } from "node:path";
+
+const PACKAGE_NAME = process.argv[2];
+const OUTPUT_PATH = resolve(process.cwd(), process.argv[3]);
```
I'm also thinking that maybe we should not let the output path be configurable.
If it's set by default than we ensure that all apps are consistent when it comes to storing their locales.
One more argument for it is that the rollup config seems to expect translations to be in `src/modules/MODULE_NAME/translations/locales/*.json` so we really should not break away from it.

luis-almeida commented on src/modules/i18n/addZendeskTranslations.ts:
```diff
@@ -0,0 +1,20 @@
+  updated_at: string;
+}
+
+export async function addZendeskTranslations(
```
Maybe calling it `localize` or `loadTranslations` would be better as `addZendeskTranslations` makes it zendesk specific and maybe confusing.

Fredx87 replied:
This function imports a file with a specific format, with the translation (key, value) map under `translations`. For this reason, I see it as "Zendesk-specific". 

Alternatively, we can put in the translation files just the translation map, like we are doing with the translation of the settings since we don't need the other info.

Another possible option is to remove this function and load the translations directly in each app (see next comment)

luis-almeida commented on src/modules/new-request-form/renderNewRequestForm.tsx:
```diff
@@ -3,12 +3,21 @@ import type { Settings } from "../garden-theme/createTheme";
+  initI18next(baseLocale);
+  await addZendeskTranslations(
+    baseLocale,
+    () => import(`./translations/locales/${baseLocale}.json`)
```
Why not pass in the json already imported instead of the dynamic import?

Fredx87 replied:
It is because the `addZendeskTranslations` function contains a try-catch around the dynamic import, to avoid throwing if the requested file is not present.

If we import the file directly here, we need to wrap it in a try-catch. At that point, the `addZendeskTranslations` is more or less useless and we can just move the logic here.

luis-almeida replied:
Got it! One can argue that it is not a traditional `fetch` call and the file should actually exist and be loaded properly for the app to work as expected.
Where I am getting at is that this might actually be one of the errors that we do want.

Totally fine to keep this behaviour though 👍🏼

Fredx87 commented on .eslintrc.js:
```diff
@@ -29,7 +29,7 @@ module.exports = {
     },
   },
-  ignorePatterns: ["assets/**"],
+  ignorePatterns: ["assets/**", "bin/**"],
```
We are linting only the `src` folder from the `lint` script: https://github.com/zendesk/copenhagen_theme/blob/ee8ef959a8e9e4a0442616bea94715ef4bc1b257/package.json#L8

I added `bin` to `ignorePatterns` because I was getting errors like `'process' is not defined` in VSCode, but this can actually be solved just by adding the comment `/* eslint-env node */` on top of the file.

 I can change that, but in any case, we are not lining the `bin` files from the CI, they will be linted only the IDE

Fredx87 commented:
I updated the PR, now the script fetches all the translation files at once. I also changed how we save the files, removing the things we don't need and storing only the mapping between the keys and the values.

luis-almeida commented on bin/update-modules-translations.mjs:
```diff
@@ -0,0 +1,51 @@
+ */
+const MODULES = {
+  "new-request-form": "new-request-form",
+};
```
I was thinking if we could eventually find a way to have this work in a "convention over configuration" type of way where it wouldn't be needed to manually keep this map. Maybe by collocating the translation within each app and check which module contains those file.
Also totally fine to keep it this way as we can always further improve.

Fredx87 replied:
> Got it! One can argue that it is not a traditional fetch call and the file should actually exist and be loaded properly for the app to work as expected.
Where I am getting at is that this might actually be one of the errors that we do want.
>
> Totally fine to keep this behaviour though 👍🏼

We don't have the guarantee that a file for a `base_locale` exists at runtime. For example, at the moment we have a `tl-ph` locale that doesn't have a `parent_identifier`, but we don't have translations for that locale.

It is probably an error in the JSON containing the mapping, but even if it gets fixed it can happen again in the future when new locales are added, or if for some reason we don't have a translation file for a locale available.

If we don't handle the error, the app will crash immediately. I think it is better to handle the error and let the app use the English strings as a fallback.

> I was thinking if we could eventually find a way to have this work in a "convention over configuration" type of way where it wouldn't be needed to manually keep this map. Maybe by collocating the translation within each app and check which module contains those file.
Also totally fine to keep it this way as we can always further improve.

It is a good point. Actually, we already have the package name in the translations YAML file. It is possible to change the script to read the package name from there, instead of manually keeping the map.
Let's merge this PR and then we can eventually revisit it

luis-almeida reviewed (approved):
Thanks for addressing the comments.
Looking pretty good 👍🏼
